File: resources/tests/HMB_Independent_Binding_Video_Slots_Regression.mjs
import assert from "node:assert/strict";
import fs from "node:fs";


async function importSource(path) {
  const source = fs.readFileSync(path, "utf8");
  return import(`data:text/javascript;base64,${Buffer.from(source).toString("base64")}`);
}


const promptPath = new URL(
  "../../widgets/HMBPromptLibraryScopedBindingWidget.js",
  import.meta.url,
);
const promptSource = fs.readFileSync(promptPath, "utf8");
const prompt = await importSource(promptPath);


const state = {
  images: [
    {
      slot: 1,
      token: "@image1",
      name: "IMAGE_01",
      present: true,
      label: "Hero",
      image_main_type: "Character",
      image_sub_type: "Full Appearance",
      source_type: "Character Appearance",
      owner: "Hero",
      scope: "Full body / full appearance",
      binding_scopes: [
        "Full body / full appearance",
        "Head / face only",
      ],
      binding_custom_scopes: ["", ""],
      binding_video_slots: [1, 3],
      marker_video: 1,
      color_picks: ["Red", "Green"],
      frame_range_enabled: false,
      frame_range_color_index: 0,
      frame_range_bindings: {},
      frame_range_binding: null,
      frame_range_selected_index: -1,
      manual: true,
    },
  ],
  videos: [
    {
      slot: 1,
      token: "@video1",
      present: true,
      label: "primary",
      source_type: "Maya Preview / Playblast",
      control_role: "Primary Unified Shot Control",
      manual: true,
    },
    {
      slot: 2,
      token: "@video2",
      present: false,
      label: "",
      source_type: "Role Required / Select Video Type",
      control_role: "",
      manual: true,
    },
    {
      slot: 3,
      token: "@video3",
      present: true,
      label: "head-guide",
      source_type: "Motion Reference",
      control_role: "Context Only",
      manual: true,
    },
  ],
  text: {},
  picker: {
    enabled: false,
    awaiting_data: false,
    suppressed: false,
    frame_metadata: [],
  },
  ui: {},
};


const normalized = prompt.normalizeState(state);
assert.deepEqual(normalized.images[0].binding_video_slots, [1, 3]);
assert.deepEqual(
  normalized.images[0].binding_scopes,
  ["Full body / full appearance", "Head / face only"],
  "Independent per-binding scopes must survive normalization.",
);
assert.equal(
  normalized.images[0].marker_video,
  1,
  "marker_video remains only the first-binding compatibility alias.",
);
assert.deepEqual(
  prompt.normalizeState(normalized).images[0].binding_video_slots,
  [1, 3],
  "Repeated widget normalization must not flatten independent video slots.",
);
assert.deepEqual(
  prompt.normalizeState(normalized).images[0].binding_scopes,
  ["Full body / full appearance", "Head / face only"],
  "Repeated widget normalization must preserve independent scopes.",
);

// The image-row video number selector exposes only video slots that currently
// contain source input. MAX_VIDEOS remains a storage limit, not a UI option
// list, and sparse slot identities must never be compacted.
assert.deepEqual(
  prompt.hmbPromptVideoBindingSlotSelection(state, 3),
  { choices: ["1", "3"], value: "3" },
  "Sparse active video slots must remain addressable by their real slot numbers.",
);
const twoActiveVideos = {
  ...state,
  videos: [
    state.videos[0],
    {
      ...state.videos[1],
      present: true,
      label: "mask",
      source_type: "Maya Preview / Playblast",
      control_role: "Mask / Guide Only",
    },
  ],
};
assert.deepEqual(
  prompt.hmbPromptVideoBindingSlotSelection(twoActiveVideos, 1),
  { choices: ["1", "2"], value: "1" },
  "Two entered videos must expose only video numbers 1 and 2.",
);
assert.deepEqual(
  prompt.hmbPromptVideoBindingSlotSelection({ ...state, videos: [state.videos[1]] }, 1),
  { choices: [""], value: "" },
  "With no entered video, the selector must not invent video numbers 1 through 10.",
);
assert.deepEqual(
  prompt.hmbPromptVideoBindingSlotSelection(state, 2),
  { choices: ["", "1", "3"], value: "" },
  "A dormant binding stays stored but must not expose its inactive number as an available choice.",
);

const customScopeState = prompt.normalizeState({
  ...state,
  images: [{
    ...state.images[0],
    image_main_type: "Custom / Context",
    image_sub_type: "Custom",
    custom_source_type: "Hero silhouette",
    scope: "Custom scope",
    binding_scopes: ["Custom scope", "Head / face only"],
    binding_custom_scopes: ["Hero silhouette", "Face detail"],
  }],
});
assert.deepEqual(customScopeState.images[0].binding_scopes, ["Custom scope", "Head / face only"]);
assert.deepEqual(
  customScopeState.images[0].binding_custom_scopes,
  ["Hero silhouette", "Face detail"],
  "Custom scope text must remain independent for every Color Pick/video binding.",
);

const longestParallelArrayState = prompt.normalizeState({
  ...state,
  images: [{
    ...state.images[0],
    color_picks: ["Red"],
    binding_scopes: ["Full body / full appearance", "Head / face only"],
    binding_custom_scopes: ["Dormant A", "Dormant B", "Dormant C"],
    binding_video_slots: [1, 3, 2],
  }],
});
assert.deepEqual(
  longestParallelArrayState.images[0].color_picks,
  ["Red", "", ""],
  "The longest of all four binding arrays determines the row count.",
);
assert.deepEqual(
  longestParallelArrayState.images[0].binding_custom_scopes,
  ["Dormant A", "Dormant B", "Dormant C"],
  "Dormant custom scopes must not be truncated when another array is shorter.",
);
assert.deepEqual(
  longestParallelArrayState.images[0].binding_video_slots,
  [1, 3, 2],
  "Dormant video-slot bindings must not be truncated when Color Pick is shorter.",
);

const nonCustomDormantState = prompt.normalizeState({
  ...state,
  images: [{
    ...state.images[0],
    binding_scopes: ["Full body / full appearance", "Head / face only"],
    binding_custom_scopes: ["Return to custom A", "Return to custom B"],
  }],
});
assert.deepEqual(
  nonCustomDormantState.images[0].binding_custom_scopes,
  ["Return to custom A", "Return to custom B"],
  "Selecting a non-Custom scope must not erase dormant Custom text.",
);

const legacyOwnerState = prompt.normalizeState({
  ...state,
  images: [{
    ...state.images[0],
    owner: "",
    binding_scopes: ["Full body / full appearance", "Handheld prop"],
  }],
});
assert.deepEqual(
  legacyOwnerState.images[0].binding_scopes,
  ["Full body / full appearance", "Handheld prop"],
);
assert.equal(
  legacyOwnerState.images[0].owner,
  "",
  "A discarded secondary Sub Type must not influence migrated image-level Target authority.",
);

const depthProvenanceState = prompt.normalizeState({
  ...state,
  videos: [
    state.videos[0],
    {
      ...state.videos[1],
      label: "generated-depth",
      present: true,
      source_type: "Depth / Spatial Reference",
      control_role: "Spatial Alignment Verification Only",
      picker_auto_label: "generated-depth",
      picker_auto_depth: {
        pair_run_id: "  pair-run-a  ",
        ignored_top_level: "must-not-round-trip",
        fields: {
          label: {
            assigned: "generated-depth",
            previous: "manual-depth",
            ignored: "must-not-round-trip",
          },
          present: { assigned: "true", previous: "false" },
          source_type: {
            assigned: "Depth / Spatial Reference",
            previous: "Custom",
          },
          custom_source_type: {
            assigned: "",
            previous: "Manual source type",
          },
          control_role: {
            assigned: "Spatial Alignment Verification Only",
            previous: "Custom Role",
          },
          custom_control_role: {
            assigned: "",
            previous: "Manual control role",
          },
          picker_auto_label: {
            assigned: "generated-depth",
            previous: "",
          },
          forbidden_field: {
            assigned: "unsafe",
            previous: "unsafe",
          },
        },
      },
      manual: true,
    },
  ],
});
const normalizedDepthProvenance =
  depthProvenanceState.videos[1].picker_auto_depth;
assert.equal(normalizedDepthProvenance.pair_run_id, "pair-run-a");
assert.deepEqual(Object.keys(normalizedDepthProvenance.fields), [
  "label",
  "present",
  "source_type",
  "custom_source_type",
  "control_role",
  "custom_control_role",
  "picker_auto_label",
]);
assert.deepEqual(normalizedDepthProvenance.fields.present, {
  assigned: true,
  previous: false,
});
assert.deepEqual(normalizedDepthProvenance.fields.label, {
  assigned: "generated-depth",
  previous: "manual-depth",
});
assert.equal(
  Object.hasOwn(normalizedDepthProvenance, "ignored_top_level"),
  false,
);
assert.equal(
  Object.hasOwn(normalizedDepthProvenance.fields, "forbidden_field"),
  false,
);
assert.deepEqual(
  prompt.normalizeState(depthProvenanceState).videos[1].picker_auto_depth,
  normalizedDepthProvenance,
  "Prompt Widget normalization must preserve sanitized Depth provenance across repeated UI round-trips.",
);


const frameImage = JSON.parse(JSON.stringify(normalized.images[0]));
frameImage.frame_range_enabled = true;
frameImage.frame_range_color_index = 1;
frameImage.frame_range_bindings = {
  "@video3::Green": {
    video_slot: "@video3",
    color_pick: "Green",
    origin: "manual",
    ranges: [{ start: 5, end: 12 }],
  },
};
frameImage.frame_range_binding = null;
frameImage.frame_range_intent = {
  version: 1,
  enabled: true,
  start_frame: 1,
  end_frame: 24,
  ranges: [{ start: 5, end: 12 }],
  selected_index: 0,
};
const frameState = {
  ...normalized,
  images: [frameImage],
  picker: {
    enabled: true,
    awaiting_data: false,
    suppressed: false,
    frame_metadata: [
      {
        video_slot: "@video3",
        fps: 24,
        start_frame: 1,
        end_frame: 24,
        frame_count: 24,
        duration_seconds: 1,
        timebase: "24/1",
        available_color_picks: ["Green"],
        origin: "maya",
        conflict: false,
        valid: true,
        warnings: [],
      },
    ],
  },
};
const frameStatus = prompt.frameRangeUiStatus(frameState, frameImage);
assert.equal(frameStatus.colorIndex, undefined);
assert.equal(frameStatus.slot, undefined);
assert.equal(frameStatus.key, undefined);
assert.equal(frameStatus.binding, undefined);
assert.deepEqual(frameStatus.ranges, [{ start: 5, end: 12 }]);
assert.equal(frameStatus.canEnable, true);


// Every mapped Color Pick row must render its own indexed video selector. A
// single row-global marker_video selector would recreate the missing-number bug.
const renderStart = promptSource.indexOf("function renderColorPickControls");
const renderEnd = promptSource.indexOf("function videoNumberOptions", renderStart);
assert.ok(renderStart >= 0 && renderEnd > renderStart);
const renderSource = promptSource.slice(renderStart, renderEnd);
assert.match(
  renderSource,
  /item\.color_picks\.map\(\(pick, pickIndex\) => `<div class="color-binding-entry"><select[\s\S]*?data-field="binding_video_slots"[\s\S]*?data-binding-index="\$\{pickIndex\}"[\s\S]*?<select[\s\S]*?data-field="color_picks"[\s\S]*?data-color-index="\$\{pickIndex\}"/,
  "Every color-binding-entry must pair an indexed video number with its indexed Color Pick.",
);
assert.doesNotMatch(
  renderSource,
  /data-field="marker_video"/,
  "The renderer must not fall back to one row-global video selector.",
);
assert.match(
  renderSource,
  /videoNumberOptions\(state, item\.binding_video_slots\[pickIndex\]\)/,
  "Initial rendering must build video-number options from current active state.",
);

const refreshStart = promptSource.indexOf("function hmbRefreshImageColorControls");
const refreshEnd = promptSource.indexOf("function hmbRefreshSourceSummaries", refreshStart);
assert.ok(refreshStart >= 0 && refreshEnd > refreshStart);
assert.match(
  promptSource.slice(refreshStart, refreshEnd),
  /hmbPromptVideoBindingSlotSelection\(state, videoSlot\)/,
  "Retained-mode refresh must use the same active-slot selector contract as initial rendering.",
);

const subtypeRenderStart = promptSource.indexOf("function renderSubtypeControls");
const subtypeRenderEnd = promptSource.indexOf("function frameRangeBarsHtml", subtypeRenderStart);
assert.ok(subtypeRenderStart >= 0 && subtypeRenderEnd > subtypeRenderStart);
const subtypeRenderSource = promptSource.slice(subtypeRenderStart, subtypeRenderEnd);
assert.match(
  subtypeRenderSource,
  /data-field="image_sub_type"/,
  "Each image row must render exactly one v2 image-level Sub Type selector.",
);
assert.doesNotMatch(
  subtypeRenderSource,
  /binding_scopes\.map/,
  "Sub Type controls must not repeat for every Color Pick/video binding.",
);

const removeStart = promptSource.indexOf('container.querySelectorAll(".remove-color-pick")');
const addStart = promptSource.indexOf('container.querySelectorAll(".add-color-pick")');
const addEnd = promptSource.indexOf('container.querySelectorAll(".add-image-source")', addStart);
assert.ok(removeStart >= 0 && addStart > removeStart && addEnd > addStart);
const removeSource = promptSource.slice(removeStart, addStart);
const addSource = promptSource.slice(addStart, addEnd);
assert.match(
  removeSource,
  /videoSlots\.pop\(\)/,
  "Removing a Color Pick must remove its parallel video number.",
);
assert.match(
  addSource,
  /videoSlots\.push\(videoSlots\[videoSlots\.length - 1\] \|\| 1\)/,
  "Adding a Color Pick must inherit the previous binding's video number.",
);
assert.match(
  addSource,
  /scopes\.push\(""\)/,
  "Adding a Color Pick must create an independent blank scope.",
);

const sourceSelectStart = promptSource.indexOf('container.querySelectorAll(".source-select")');
const sourceSelectEnd = promptSource.indexOf('container.querySelectorAll(".move-image-up', sourceSelectStart);
const sourceSelectSource = promptSource.slice(sourceSelectStart, sourceSelectEnd);
assert.doesNotMatch(
  sourceSelectSource,
  /binding_custom_scopes\[bindingIndex\]\s*=/,
  "Changing a scope selector must not erase or rewrite dormant Custom text.",
);


console.log("HMB independent binding_video_slots widget regression: PASS");

File: resources/tests/HMB_Independent_Binding_Video_Slots_Regression.py
from __future__ import annotations

import copy
import importlib.util
import json
import sys
from pathlib import Path


ROOT = Path(__file__).resolve().parents[2]


def load(name: str):
    path = ROOT / f"{name}.py"
    spec = importlib.util.spec_from_file_location(name, path)
    module = importlib.util.module_from_spec(spec)
    sys.modules[name] = module
    spec.loader.exec_module(module)
    return module


prompt = load("HMBPromptLibrary")
agent = load("HMBAgentLibrary")


def prompt_json_section(payload: str, header: str):
    lines = payload.splitlines()
    assert lines[0] == "HMB_GP_Production"
    assert header in lines
    return json.loads(lines[lines.index(header) + 1])


def active_video(slot: int, label: str, source_type: str, control_role: str):
    item = prompt._default_video_item(slot)
    item.update(
        {
            "present": True,
            "label": label,
            "source_type": source_type,
            "control_role": control_role,
            "video_main_type": (
                "Maya Preview / Playblast"
                if source_type == "Maya Preview / Playblast"
                else "Motion Reference"
            ),
            "video_sub_type": (
                "Original Preview"
                if source_type == "Maya Preview / Playblast"
                else "Local Motion"
            ),
            "manual": True,
        }
    )
    return item


# One image may bind independent Color Pick rows to different videos. Repeated
# normalization preserves every authored scope/custom scope and slot; the
# image-level Main/Sub pair does not rewrite those relationship fields.
state = prompt._default_widget_state()
state["images"] = [prompt._default_image_item(1)]
state["images"][0].update(
    {
        "present": True,
        "label": "Hero",
        "image_main_type": "Character",
        "image_sub_type": "Full Appearance",
        "source_type": "Character Appearance",
        "owner": "Hero",
        "binding_scopes": [
            "Full body / full appearance",
            "Head / face only",
        ],
        "binding_custom_scopes": ["", ""],
        "binding_video_slots": [1, 3],
        "marker_video": 1,
        "color_picks": ["Red", "Green"],
    }
)
state["videos"] = [
    active_video(
        1,
        "primary",
        "Maya Preview / Playblast",
        "Primary Unified Shot Control",
    ),
    prompt._default_video_item(2),
    active_video(3, "head-guide", "Motion Reference", "Context Only"),
]

normalized = prompt._normalize_state(state)
image = normalized["images"][0]
assert image["binding_video_slots"] == [1, 3]
assert image["marker_video"] == 1
assert image["binding_scopes"] == [
    "Full body / full appearance",
    "Head / face only",
]
assert prompt._normalize_state(copy.deepcopy(normalized)) == normalized

custom_scope_image = prompt._default_image_item(1)
custom_scope_image.update(
    {
        "image_main_type": "Custom / Context",
        "image_sub_type": "Custom",
        "custom_source_type": "Hero custom binding",
        "color_picks": ["Red", "Green"],
        "binding_scopes": ["Custom scope", "Head / face only"],
        "binding_custom_scopes": ["Hero silhouette", "Face detail"],
        "binding_video_slots": [1, 3],
    }
)
prompt._normalize_image_binding_fields(custom_scope_image)
assert custom_scope_image["binding_scopes"] == ["Custom scope", "Head / face only"]
assert custom_scope_image["binding_custom_scopes"] == [
    "Hero silhouette",
    "Face detail",
]

legacy_owner_state = prompt._default_widget_state()
legacy_owner_state["images"] = [
    {
        "slot": 1,
        "present": True,
        "label": "Hero",
        "image_main_type": "Character",
        "image_sub_type": "Full Appearance",
        "source_type": "Character Appearance",
        "owner": "",
        "color_picks": ["Red", "Green"],
        "binding_scopes": [
            "Full body / full appearance",
            "Handheld prop",
        ],
    }
]
legacy_owner_image = prompt._normalize_state(legacy_owner_state)["images"][0]
assert legacy_owner_image["binding_scopes"] == [
    "Full body / full appearance",
    "Handheld prop",
]
assert legacy_owner_image["owner"] == ""

entries = prompt._image_binding_entries(image)
assert [
    (entry["marker_video"], entry["color"], entry["scope"])
    for entry in entries
] == [
    (1, "Red", "Full body / full appearance"),
    (3, "Green", "Head / face only"),
]

compiled = prompt._build_data_only_prompt_package(normalized)
compiled_job = prompt_json_section(compiled, "HMB JOB DATA (JSON):")
assert compiled_job["images"][0]["bindings"] == [
    {
        "video": "@video1",
        "marker_color": "Red",
        "target_scope": "Full body / full appearance",
    },
    {
        "video": "@video3",
        "marker_color": "Green",
        "target_scope": "Head / face only",
    },
]
assert compiled_job["images"][0]["image_main_type"] == "Character"
assert compiled_job["images"][0]["image_sub_type"] == "Full Appearance"


# The selected Color Pick index owns the Range address. Index 1 must resolve to
# binding_video_slots[1] (@video3), not to the compatibility marker_video value.
range_state = copy.deepcopy(normalized)
range_image = range_state["images"][0]
range_image["frame_range_intent"] = {
    "version": 1,
    "enabled": True,
    "start_frame": 1,
    "end_frame": 24,
    "ranges": [{"start": 5, "end": 12}],
    "selected_index": 0,
}
range_image["frame_range_enabled"] = True
range_image["frame_range_color_index"] = 1
range_image["frame_range_bindings"] = {
    "@video3::Green": {
        "video_slot": "@video3",
        "color_pick": "Green",
        "origin": "manual",
        "ranges": [{"start": 5, "end": 12}],
    }
}
range_state["picker"] = {
    "enabled": True,
    "awaiting_data": False,
    "suppressed": False,
    "run_id": "independent-binding-video-range",
    "scene": "C:/shots/independent_binding.mb",
    "video_path": "",
    "camera": "|shotCam",
    "markers": [],
    "matched_images": 0,
    "frame_metadata": [
        {
            "video_slot": "@video3",
            "fps": 24,
            "start_frame": 1,
            "end_frame": 24,
            "frame_count": 24,
            "duration_seconds": 1,
            "timebase": "24/1",
            "available_color_picks": ["Green"],
            "origin": "maya",
            "conflict": False,
            "valid": True,
            "warnings": [],
        }
    ],
}
range_state = prompt._normalize_state(range_state)
range_image = range_state["images"][0]
current_binding = prompt._current_frame_range_binding(range_image)
assert current_binding is not None
assert current_binding["video_slot"] == "@video3"
assert current_binding["color_pick"] == "Green"
assert current_binding["ranges"] == [{"start": 5, "end": 12}]

range_compiled = prompt._build_data_only_prompt_package(range_state)
assert prompt_json_section(
    range_compiled,
    "HMB JOB DATA (JSON):",
)["frame_ranges"] == [{
    "image": "@image1",
    "video": "@video3",
    "marker_color": "Green",
    "enabled": True,
    "origin": "manual",
    "domain": {"start_frame": 1, "end_frame": 24},
    "segments": [{"start_frame": 5, "end_frame": 12}],
}]


# Deactivating @video3 makes its binding dormant without deleting user intent.
# The @video1 binding, shared scope, and optional frame instruction remain intact
# so the independent @video3 source can return later.
inactive_state = copy.deepcopy(normalized)
inactive_state["images"][0]["frame_range_intent"] = {
    "version": 1,
    "enabled": True,
    "start_frame": 1,
    "end_frame": 24,
    "ranges": [{"start": 5, "end": 12}],
    "selected_index": 0,
}
inactive_state["images"][0]["frame_range_enabled"] = True
inactive_state["images"][0]["frame_range_color_index"] = 1
inactive_state["images"][0]["frame_range_bindings"] = {
    "@video3::Green": {
        "video_slot": "@video3",
        "color_pick": "Green",
        "origin": "manual",
        "ranges": [{"start": 5, "end": 12}],
    }
}
inactive_state["videos"][2].update(
    {
        "present": False,
        "label": "",
        "source_type": "Ignore / Unused",
        "control_role": "",
    }
)
inactive = prompt._normalize_state(inactive_state)["images"][0]
assert inactive["binding_video_slots"] == [1, 3]
assert inactive["color_picks"] == ["Red", "Green"]
assert "@video3::Green" in inactive["frame_range_bindings"]
assert inactive["frame_range_color_index"] == 1
assert inactive["binding_scopes"] == [
    "Full body / full appearance",
    "Head / face only",
]
assert inactive["marker_video"] == 1


# Picker re-sync replaces only its exact automatic binding and preserves a
# user-added binding with its independent video number and shared image subtype.
def picker_payload(color: str, run_id: str):
    return {
        "mode": "maya",
        "run_id": run_id,
        "active_slot_count": 3,
        "videos": [
            {"video_slot": 1, "video_path": "C:/shots/primary.mp4"},
            {"video_slot": 2, "video_path": "C:/shots/mask.mp4"},
            {"video_slot": 3, "video_path": "C:/shots/detail.mp4"},
        ],
        "markers": [
            {
                "asset_id": "Hero",
                "color": color,
                "video_slot": 2,
                "picker_order": 1,
            }
        ],
    }


picker_state = prompt._default_widget_state()
picker_state["images"][0].update(
    {
        "present": True,
        "label": "Hero",
        "image_main_type": "Character",
        "image_sub_type": "Full Appearance",
        "source_type": "Character Appearance",
        "owner": "Hero",
        "binding_scopes": ["Full body / full appearance"],
        "binding_custom_scopes": [""],
    }
)
picker_state = prompt._apply_picker_payload(
    picker_state,
    picker_payload("Red", "independent-auto-1"),
    connected=True,
)
picker_image = picker_state["images"][0]
picker_image["color_picks"] = ["Red", "Yellow"]
picker_image["binding_video_slots"] = [2, 3]
picker_image["binding_scopes"] = [
    "Full body / full appearance",
    "Head / face only",
]
picker_image["binding_custom_scopes"] = ["", ""]
picker_image["marker_video"] = 2

picker_state = prompt._apply_picker_payload(
    picker_state,
    picker_payload("Green", "independent-auto-2"),
    connected=True,
)
picker_image = picker_state["images"][0]
assert picker_image["color_picks"] == ["Green", "Yellow"]
assert picker_image["binding_video_slots"] == [2, 3]
assert picker_image["binding_scopes"] == [
    "Full body / full appearance",
    "Head / face only",
]
assert picker_image["picker_auto_color"] == "Green"
assert picker_image["picker_auto_video"] == 2


print("HMB independent binding_video_slots Python regression: PASS")
